### PR TITLE
Feature/middleware decorator

### LINKED ISSUE #41: Security decorators (middleware) ?
Hey, great library.

I was wondering if its possible to pass in custom middleware on a route or is it planned ?

or even better, protecting routes using security decorators ?

### PR DESCRIPTION
**Pull request by buehler**
Closes #41 
- Adds open param list to all routes and controller decorator
- Shows a simple example with http basic auth (passportJS)

### CODE REVIEW

File: examples/simple-middleware/app.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @Put()
+    public createDemo(@Body({required: true}) body: Demo): void {
+        this.demos.push(body);
+    }
+
+    @Post(':id')
+    public updateDemo(@UrlParam('id') id: number, @Body({required: true}) body: Demo): void {
+        let filtered = this.demos.filter(d => d.id === id);
+        if (!filtered.length) {
+            throw new Error('not found.');
```
Comment (petermanser) [inline examples/simple-middleware/app.ts]: shouldn't this be a 404?

File: controllers/Middleware.spec.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+describe('Middleware', () => {
+
+    afterEach(() => {
+        resetControllerRegistrations();
+    });
+
+    it('should call controller middleware', () => {
+        let spy = sinon.spy();
+
+        @Controller('test', spy)
```
Comment (petermanser) [inline controllers/Middleware.spec.ts:56]: add a second route, to test if it's called for each route.

File: controllers/Middleware.spec.ts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        class Ctrl {
+
+            @Route()
```
Comment (petermanser) [inline controllers/Middleware.spec.ts]: add a test if `@Controller` and `@Route` register middlewares

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,9 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ## [Unreleased]
 ### Added
 - isArray validator can now validate items with validators and multiple types.
+- `@Controller` and all route decorators do accept `RequestHandler` as middlewares.
```
Comment (petermanser) [inline CHANGELOG.md]: `@Controller` and `@Route` decorators optionally accept `RequestHandler` as middlewares.

File: controllers/Middleware.spec.ts (hunk 2 above)
Comment (buehler) [inline controllers/Middleware.spec.ts:56]: 5f6b42d2ae4b1500a24145bc853e74b8521e4689

File: CHANGELOG.md (hunk 4 above)
Comment (buehler) [inline CHANGELOG.md]: 5f6b42d2ae4b1500a24145bc853e74b8521e4689

**Pull request merged**